README.md
[![Build Status](https://travis-ci.org/KlausVigo/ggnetworx.svg?branch=master)](https://travis-ci.org/KlausVigo/ggnetworx)
[![codecov](https://codecov.io/gh/KlausVigo/ggnetworx/branch/master/graph/badge.svg)](https://codecov.io/gh/KlausVigo/ggnetworx)
[![Coverage Status](https://coveralls.io/repos/github/KlausVigo/ggnetworx/badge.svg?branch=master)](https://coveralls.io/github/KlausVigo/ggnetworx?branch=master)




ggnetworx
========================================================

ggnetworx provides function to plot phylogenetic networks of class `evonet` from *ape* and split graphs of class `networx` from *phangorn*.

You can install `remotes::install_github("KlausVigo/ggnetworx")`

To install *ggnetworx* you may need to install the *ggtree* package from bioconductor first:
```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ggtree")
```

If you use ggnetworx please cite:

Paradis E. & Schliep K. 2018. ape 5.0: an environment for modern phylogenetics 
and evolutionary analyses in R. Bioinformatics 35: 526-528.

Schliep K.P. 2011. phangorn: phylogenetic analysis in R. Bioinformatics, 27(4) 592-593

Schliep, K., Potts, A. J., Morrison, D. A. and Grimm, G. W. (2017), Intertwining phylogenetic trees and networks. Methods Ecol Evol. 8, 1212–1220.

and also 

Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36


License
-------
phangorn is licensed under the AGPL v3.


NEWS
CHANGES IN VERSION 0.99
-------------------------

NEW FEATURES

    o initial release. The main functions are ggsplitnet and ggevonet to
    visualize split (or implicit) networks (unrooted, undirected) and explicit
    networks (rooted, directed) with reticulations.


tests/testthat.R
library(testthat)
suppressPackageStartupMessages(library(ggnetworx))

test_check("ggnetworx")


tests/testthat/test_networx.R
context("networx")


data(yeast, package="phangorn")
dm <- phangorn::dist.ml(yeast)
nnet <- phangorn::neighborNet(dm)

(enet <- ape::read.evonet(text="((a:2,(b:1)#H1:1):1,(#H1,c:1):2);"))

splits_network <-  ggsplitnet(nnet) + geom_tiplab2()
reticulate_network <- ggevonet(enet) + geom_tiplab()


# Visual tests ------------------------------------------------------------
#test_that("visual appearance", {
#    testthat::skip_on_cran()

#    vdiffr::expect_doppelganger("Basic splits network", splits_network)
#    vdiffr::expect_doppelganger("Basic reticulate network", reticulate_network)
#})


R/node_depth.R
#' These functions return the depths or heights of nodes and tips.
#'
#' @title Depth of Nodes
#' @param x an object of class "evonet"
#' @param \dots Further arguments passed to or from other methods.
#' @return a vector with the depth of the nodes
#' @seealso \code{\link[ape]{node.depth}}
#' @examples
#' z <- ape::read.evonet(text = "((1,((2,(3,(4)Y#H1)g)e,
#' (((Y#H1, 5)h,6)f)X#H2)c)a,((X#H2,7)d,8)b)r;")
#' nd <- node.depth.evonet(z)
#' z$edge.length <- nd[z$edge[,1]] - nd[z$edge[,2]]
#' ggevonet(z)
#'
#' @importFrom phangorn getRoot Ancestors Descendants
#' @export
node.depth.evonet <- function(x, ...){
   x <- ape::reorder.phylo(x)
   root <- getRoot(x)
   max_nodes <- max(x$edge)
   nTip <- Ntip(x)
   desc <- Descendants(x, seq_len(max_nodes), "children")
   anc <- Ancestors(x)
   pa <- vector("list", max_nodes)
   ind <- which(x$edge[,2] > Ntip(x))
   pa[x$edge[ind,2]] <- x$edge[ind,1]
   for(i in seq_len(nrow(x$reticulation))){
      pa[[x$reticulation[i,2]]] <- sort( c(pa[[x$reticulation[i,2]]],
                                           x$reticulation[i,1] ) )
#       pa[[x$reticulation[i,1] ]] <- numeric(0)
   }
   ind <- which(lengths(pa) > 0)
   depth <- numeric(max_nodes)
   depth[root] <- 1
   done <- logical(max_nodes)
   done[root] <- TRUE
   candidates <- desc[[root]]
   candidates <- candidates[candidates>nTip]
   d <- 1
   while(length(candidates)>0){
      active <- vapply(candidates, function(x) all(done[pa[[x]]]), FALSE)
      tmp <- which(active)[1]  #sample(active,1)
      candidates <- c(candidates, desc[[candidates[tmp] ]])
      candidates <- candidates[candidates>nTip]
      d <- d+1
      done[candidates[tmp]] <- TRUE
      depth[candidates[tmp]] <- d
      candidates <- candidates[-tmp]
   }
   depth <- d+2 - depth
   depth[seq_len(nTip)] <- 1
   depth
}


R/minimize_overlap.R
#' @title minimize_overlap
#' reduces reticulation lines crossing over in plots
#' @param x Tree of class 'evonet'
#' @return  A Tree with rotated nodes of class 'evonet'
#' @importFrom ape node.height rotate Ntip
#' @importFrom phangorn Ancestors
#' @author L. Francisco Henao Diaz
#' @examples
#' fishnet <- ape::read.evonet(text="(Xalvarezi,Xmayae,((Xsignum,((Xmonticolus,(Xclemenciae_F2,#H25:::0.032485577051040423):0.979116879342896):1.4858786525152954,(((((((((Xgordoni,Xmeyeri):0.26111415598145865,Xcouchianus):3.5325554148283564,Xvariatus):0.6445452395880061,Xevelynae):0.41219807813960146,(Xxiphidium,#H24:0.0::0.16538333427567822):1.4020042304402076):0.2921320715527958,Xmilleri):0.4688709909949548,Xandersi):0.6492623942884177,Xmaculatus):1.0200542801748702,(((Xmontezumae,(Xcortezi,(Xbirchmanni_GARC,Xmalinche_CHIC2):0.8943004029517725):0.6903622798656042):0.4926275928330048,((Xnigrensis,Xmultilineatus):1.4508210422790195,(Xpygmaeus,Xcontinens):1.9686498919562234):2.428065080719157):2.0435103751744865)#H24:0.6476450483433508::0.8346166657243218):0.7816554661308311):1.9841381634366537):0.31614675520900004,(Xhellerii)#H25:::0.9675144229489596):0.28226883225049243);")
#' fishnet$edge.length <- NULL
#' new_tre <- minimize_overlap(fishnet)
#'
#' par(mfrow=c(1,2))
#' ggevonet(fishnet)
#' ggevonet(new_tre)
#'
#' net2 <- ape::read.evonet(text="(15,(1,((14,(#H1,(((12,13),(11,#H3)),(7,
#'    ((10)#H3,(8,9)))))),((((2,3))#H2,(6,(5,(#H2,4)))))#H1)));")
#' # Cui et al. 2013 Evol.
#' new_net2 <- minimize_overlap(net2)
#' ggevonet(net2)
#' ggevonet(new_net2)
#'
#' @export
minimize_overlap <- function(x){
    if(class(x)[1]!='evonet') stop("x should be an 'evonet' class")
    n_iter <- round(x$Nnode*3/4)
#    r_hist <- c()
    for(j in seq_len(n_iter)){
        h <- node.height(x)
        best_r <- sum(abs(h[x$reticulation[,1]]- h[x$reticulation[,2]]))
        best_c <- -1
#        r_hist[j] <- best_r
        nodes2rot <- intersect(sort(unique(unlist(Ancestors(x,
                        c(x$reticulation))))), which(tabulate(x$edge[,1]) > 1) )
        for(i in seq_along(nodes2rot)){
            nh <- node.height(ape::rotate(x, nodes2rot[i]))
            best_nr <- sum(abs(nh[x$reticulation[,1]] - nh[x$reticulation[,2]]))
            if(best_nr < best_r){
                best_c <- nodes2rot[i]
                best_r <- best_nr
            }
        }
        if(best_c > 0) x <- ape::rotate(x, best_c)
        else break()
    }
    x
}



R/ggnetworx.R
#' @importFrom magrittr %>%
#' @export
magrittr::'%>%'

#' @method fortify evonet
#' @importFrom ggplot2 fortify
#' @export
fortify.evonet <- function(model, data, layout="rectangular", ladderize=FALSE,
                    right=FALSE, mrsd=NULL, as.Date  =FALSE, ...){
    class(model) <- "phylo"
#  ggtree:::fortify.phylo
    df <- fortify(model, ladderize=ladderize)

    hybridEdge <- logical(nrow(df))
    hybridEdge[grep("#", df$label)] <- TRUE
    df <- cbind(df, hybridEdge=hybridEdge)

    reticulation <- model$reticulation
    df.ret <- df[reticulation[,1], , drop=FALSE]
#    df.ret <- df[reticulation[,2], , drop=FALSE]
    df.ret[,c("node", "parent")] <- reticulation
    df.ret[, "hybridEdge"] <- TRUE
    df <- rbind(df, df.ret)
    df
}


#' drawing phylogenetic tree from phylo object
#'
#'
#' @title ggevonet
#' @param tr phylo object
#' @param mapping aes mapping
#' @param layout one of 'rectangular', 'slanted'
#' @param open.angle open angle, only for 'fan' layout
#' @param mrsd most recent sampling date
#' @param as.Date logical whether using Date class in time tree
#' @param yscale y scale
#' @param yscale_mapping yscale mapping for category variable
#' @param ladderize logical
#' @param right logical
#' @param branch.length variable for scaling branch, if 'none' draw cladogram
#' @param ndigits number of digits to round numerical annotation variable
#' @param min_crossing logical, rotate clades to minimize crossings
#' @param ... additional parameter
#' @return tree
#' @seealso \code{\link[ape]{evonet}}, \code{\link[ggtree]{ggtree}}
#' @importFrom ggplot2 ggplot
#' @importFrom ggplot2 xlab
#' @importFrom ggplot2 ylab
#' @importFrom ggplot2 annotate
#' @importFrom ggplot2 scale_x_reverse
#' @importFrom ggplot2 ylim
#' @importFrom ggplot2 coord_flip
#' @importFrom ggplot2 coord_polar
#' @importFrom ggplot2 aes
#' @importFrom ggplot2 aes_
#' @importFrom ggtree geom_tree2
#' @importFrom ggtree theme_tree
#' @importFrom phangorn coords
#' @importFrom ape read.evonet
#' @author Klaus Schliep
#' @examples
#' (enet <- ape::read.evonet(text="((a:2,(b:1)#H1:1):1,(#H1,c:1):2);"))
#' ggevonet(enet) + geom_tiplab()
#' @export
ggevonet <- function (tr, mapping=NULL, layout="slanted", open.angle=0,
            mrsd=NULL, as.Date=FALSE, yscale="none", yscale_mapping=NULL,
            ladderize=FALSE, right=FALSE, branch.length="branch.length",
            ndigits=NULL, min_crossing=TRUE, ...)
{
    layout <- match.arg(layout, c("rectangular", "slanted"))
# , "fan", "circular", "radial", "unrooted", "equal_angle", "daylight"

    if(is.null(tr$edge.length)){
        nh <- node.depth.evonet(tr)
        tr$edge.length <- nh[tr$edge[,1]] - nh[tr$edge[,2]]
    }
    if(min_crossing){
        tr <- minimize_overlap(tr)
    }
    if (yscale != "none") {
        layout <- "slanted"
    }
    if (is.null(mapping)) {
        mapping <- aes_(~x, ~y)
    }
    else {
        mapping <- modifyList(aes_(~x, ~y), mapping)
    }
    mapping <- modifyList(aes(linetype = hybridEdge), mapping)
    p <- ggplot(tr, mapping=mapping, layout=layout, mrsd=mrsd, as.Date=as.Date,
                yscale=yscale, yscale_mapping=yscale_mapping,
                ladderize=ladderize, right=right, branch.length=branch.length,
                ndigits=ndigits, ...)

    p <- p + geom_tree2(layout=layout, ...)
    p <- p + theme_tree(legend.position="none")
    class(p) <- c("ggtree", class(p))
    return(p)
}


#' @method fortify networx
#' @importFrom phangorn getRoot
#' @export
fortify.networx <- function(model, data, layout="unrooted", ladderize=FALSE,
                            right=FALSE, mrsd=NULL, as.Date=FALSE, ...){
##    root <- getRoot(model)
    nTips <- length(model$tip.label)
    label <- character(nrow(model$edge))
    isTip <- logical(nrow(model$edge))  ## edge leading to tip
    if(!is.null(model$translate)){
        ind <- match(model$translate$node, model$edge[,2])
        label[ind] <- model$translate$label
    }
    else{
        ind <- match(seq_len(nTips), model$edge[,2])
        label[ind] <- model$tip.label
    }
    isTip[ind] <- TRUE
    df <- data.frame(node=model$edge[,2], parent=model$edge[,1],
                    branch.length=model$edge.length, split=model$splitIndex,
                    label=label, isTip=isTip)
    if(!is.null(model$.plot)) coord <- model$.plot$vertices
    else coord <- coords(model, dim="2D")
    df <- cbind(df, x=coord[df$node,1], y=coord[df$node,2],
                xend=coord[df$parent,1], yend=coord[df$parent,2])
    angle <- atan2(df$y - df$yend, df$x - df$xend) * 360 / (2*pi)
    angle[angle<0] <- angle[angle<0] + 360
    df <- cbind(df, angle=angle)
    df
}


#' drawing phylogenetic tree from phylo object
#'
#'
#' @title ggsplitnet
#' @param tr phylo object
#' @param mapping aes mapping
#' @param layout one of 'rectangular', 'slanted', 'fan', 'circular', 'radial',
#' 'equal_angle' or 'daylight'
#' @param open.angle open angle, only for 'fan' layout
#' @param mrsd most recent sampling date
#' @param as.Date logical whether using Date class in time tree
#' @param yscale y scale
#' @param yscale_mapping yscale mapping for category variable
#' @param ladderize logical
#' @param right logical
#' @param branch.length variable for scaling branch, if 'none' draw cladogram
#' @param ndigits number of digits to round numerical annotation variable
#' @param ... additional parameter
#' @return tree
#' @seealso \code{\link[ggtree]{ggtree}}, \code{\link[phangorn]{networx}},
#' \code{\link[phangorn]{consensusNet}}, \code{\link[phangorn]{neighborNet}}
#' @references Schliep, K., Potts, A. J., Morrison, D. A. and Grimm, G. W.
#' (2017), Intertwining phylogenetic trees and networks.
#' \emph{Methods Ecol Evol}. \bold{8}, 1212--1220. doi:10.1111/2041-210X.12760
#' @importFrom utils modifyList
#' @importFrom ggplot2 ggplot
#' @importFrom ggplot2 xlab
#' @importFrom ggplot2 ylab
#' @importFrom ggplot2 annotate
#' @importFrom ggplot2 scale_x_reverse
#' @importFrom ggplot2 ylim
#' @importFrom ggplot2 coord_flip
#' @importFrom ggplot2 coord_polar
#' @importFrom ggtree theme_tree
#' @author Klaus Schliep
#' @examples
#' data(yeast, package="phangorn")
#' dm <- phangorn::dist.ml(yeast)
#' nnet <- phangorn::neighborNet(dm)
#' ggsplitnet(nnet) + geom_tiplab2()
#'
#'
#' @export
ggsplitnet <- function (tr, mapping=NULL, layout="slanted", open.angle=0,
        mrsd=NULL, as.Date=FALSE, yscale="none", yscale_mapping=NULL,
        ladderize=FALSE, right=FALSE, branch.length="branch.length",
        ndigits=NULL, ...)
{
    layout <- match.arg(layout, c("slanted"))
# "rectangular", "fan", "circular", "radial", "unrooted", "equal_angle",
#    "daylight"
    if (is.null(mapping)) {
        mapping <- aes_(~x, ~y)
    }
    else {
        mapping <- modifyList(aes_(~x, ~y), mapping)
    }
    p <- ggplot(tr, mapping=mapping, layout=layout, mrsd=mrsd, as.Date=as.Date,
                yscale=yscale, yscale_mapping=yscale_mapping,
                ladderize=ladderize, right=right, branch.length=branch.length,
                ndigits=ndigits, ...)

    p <- p + geom_splitnet(layout=layout, ...)
    p <- p + theme_tree()
    class(p) <- c("ggtree", class(p))
    return(p)
}





##' add tree layer
##'
##'
##' @title geom_splitnet
##' @param layout one of 'rectangular', 'slanted', 'circular', 'radial' or
##' 'unrooted'
##' @param ... additional parameter
##' @return tree layer
##' @examples
##' data(yeast, package="phangorn")
##' dm <- phangorn::dist.ml(yeast)
##' nnet <- phangorn::neighborNet(dm)
##' ggplot(nnet, aes(x, y))  + geom_splitnet() + theme_tree()
##' @importFrom ggplot2 geom_segment
##' @importFrom ggplot2 aes
##' @export
##' @author Klaus Schliep
geom_splitnet <- function(layout="slanted", ...) {
    x <- y <- xend <- yend <- parent <- NULL
    lineend <- "round"
    if (layout == "rectangular" || layout == "fan" || layout == "circular") {
        list(
            geom_segment(aes(x=x, xend=xend, y=y, yend=y),
                        lineend=lineend, ...),
            geom_segment(aes(x=xend, xend=xend, y=y, yend=yend),
                        lineend=lineend, ...)
        )
    } else if (layout == "slanted" || layout == "radial" ||
                layout == "unrooted") {
        geom_segment(aes(x=x, xend=xend, y=y, yend=yend),
                    lineend  = lineend, ...)
    }
}
